src/Microsoft.Health.Fhir.Shared.Api/Features/ActionResults/OperationVersionsResult.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using EnsureThat;
+using Hl7.Fhir.Model;
+using Hl7.Fhir.Rest;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Health.Fhir.Core.Features.Operations.Versions;
+using Microsoft.Net.Http.Headers;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Health.Fhir.Api.Features.ActionResults
+{
+    /// <summary>
+    /// Used to return the result of a versions operation.
+    /// </summary>
+    public class OperationVersionsResult : BaseActionResult<VersionsResult>
+    {
+        private readonly VersionsResult _versionsResult;
+
+        private HttpContext _httpContext;
+        private IList<MediaTypeHeaderValue> _acceptHeaders;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OperationVersionsResult"/> class.
+        /// </summary>
+        /// <param name="versionsResult">The result for the versions operation.</param>
+        /// <param name="statusCode">The HTTP status code.</param>
+        public OperationVersionsResult(VersionsResult versionsResult, HttpStatusCode statusCode)
+            : base(versionsResult, statusCode)
+        {
+            EnsureArg.IsNotNull(versionsResult, nameof(versionsResult));
+            _versionsResult = versionsResult;
+        }
+
+        public override Task ExecuteResultAsync(ActionContext context)
+        {
+            _httpContext = context.HttpContext;
+            _acceptHeaders = _httpContext.Request.GetTypedHeaders().Accept;
+            return base.ExecuteResultAsync(context);
+        }
+
+        protected override object GetResultToSerialize()
+        {
+            if (_acceptHeaders != null && _acceptHeaders.All(a => a.MediaType != "*/*"))
+            {
+                foreach (MediaTypeHeaderValue acceptHeader in _acceptHeaders)
+                {
+                    // If the accept header is application/[json/xml]
+                    if (string.Compare(acceptHeader.SubType.ToString(), ContentType.FORMAT_PARAM_XML, StringComparison.OrdinalIgnoreCase) == 0 ||
+                        string.Compare(acceptHeader.SubType.ToString(), ContentType.FORMAT_PARAM_JSON, StringComparison.OrdinalIgnoreCase) == 0)
+                    {
+                        // Follow the format outlined in the spec: https://www.hl7.org/fhir/capabilitystatement-operation-versions.html.
+                        return _versionsResult;
+                    }
+
+                    // If the accept header is application/fhir+[json/xml]
+                    if (string.Equals("fhir", acceptHeader.Copy().SubTypeWithoutSuffix.ToString(), StringComparison.OrdinalIgnoreCase))
```

Comment:
I think I had `.copy` in there before as a typo, shouldn't be required, sorry for the confusion